drivers/modem/modem_cellular.c

```diff
@@ -717,8 +726,67 @@ static int modem_cellular_on_await_power_on_state_enter(struct modem_cellular_da
 static void modem_cellular_await_power_on_event_handler(struct modem_cellular_data *data,
 							enum modem_cellular_event evt)
 {
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
 	switch (evt) {
 	case MODEM_CELLULAR_EVENT_TIMEOUT:
+		if (config->set_baudrate_chat_script != NULL) {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_SET_BAUDRATE);
+		}
+		else {
+			modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_RUN_INIT_SCRIPT);
+		}
+
+	break;
+
+	case MODEM_CELLULAR_EVENT_SUSPEND:
+		modem_cellular_enter_state(data, MODEM_CELLULAR_STATE_IDLE);
+		break;
+
+	default:
+		break;
+	}
+}
+
+static int modem_cellular_on_set_baudrate_state_enter(struct modem_cellular_data *data)
+{
+	modem_pipe_attach(data->uart_pipe, modem_cellular_bus_pipe_handler, data);
+	return modem_pipe_open_async(data->uart_pipe);
+}
+
+static void modem_cellular_set_baudrate_event_handler(struct modem_cellular_data *data,
+							 enum modem_cellular_event evt)
+{
+	const struct modem_cellular_config *config =
+		(const struct modem_cellular_config *)data->dev->config;
+
+	switch (evt) {
+	case MODEM_CELLULAR_EVENT_BUS_OPENED:
+		modem_chat_attach(&data->chat, data->uart_pipe);
+		modem_chat_run_script_async(&data->chat, config->set_baudrate_chat_script);
+		break;
+
+	case MODEM_CELLULAR_EVENT_SCRIPT_SUCCESS:
+		/* Let modem reconfigure */
+		k_msleep(CONFIG_MODEM_CELLULAR_NEW_BAUDRATE_DELAY);
+		__fallthrough;
```

Comment:
Delays er performed using modem_cellular_start_timer() and waiting for the `MODEM_CELLULAR_EVENT_TIMEOUT`, sleeping is not allowed in this driver as it blocks the system work queue

Reply:
Makes sense